group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/AbstractUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,256 @@
+ */
+package org.apache.kafka.coordinator.group.assignor;
+
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.server.common.TopicIdPartition;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * The assignment builder is used to construct the target assignment based on the members' subscriptions.
+ *
+ * This class contains common utility methods and a class for obtaining and storing rack information.
+ */
+public abstract class AbstractUniformAssignmentBuilder {
+    protected abstract GroupAssignment buildAssignment();
+
+    /**
+     * Determines if rack-aware assignment is appropriate based on the provided rack information.
+     *
+     * @param memberRacks               Racks where members are located.
+     * @param allPartitionRacks         Racks where partitions are located.
+     * @param racksPerPartition         Map of partitions to their associated racks.
+     *
+     * @return {@code true} if rack-aware assignment should be applied; {@code false} otherwise.
+     */
+    protected static boolean useRackAwareAssignment(
+        Set<String> memberRacks,
+        Set<String> allPartitionRacks,
+        Map<TopicIdPartition, Set<String>> racksPerPartition
+    ) {
+        if (memberRacks.isEmpty() || Collections.disjoint(memberRacks, allPartitionRacks))
+            return false;
+        else {
+            return !racksPerPartition.values().stream().allMatch(allPartitionRacks::equals);
+        }
+    }
+
+    /**
+     * Adds the topic's partition to the member's target assignment.
+     */
+    protected static void addPartitionToAssignment(
+        Map<String, MemberAssignment> memberAssignments,
+        String memberId,
+        Uuid topicId,
+        int partition
+    ) {
+        memberAssignments.get(memberId)
+            .targetPartitions()
+            .computeIfAbsent(topicId, __ -> new HashSet<>())
+            .add(partition);
+    }
+
+    /**
+     * Constructs a set of {@code TopicIdPartition} including all the given topic Ids based on their partition counts.
+     *
+     * @param topicIds                      The topic Ids.
+     * @param subscribedTopicDescriber      Utility to fetch the partition count for a given topic.
+     *
+     * @return Set of {@code TopicIdPartition} including all the provided topic Ids.
+     */
+    protected static Set<TopicIdPartition> topicIdPartitions(
+        Collection<Uuid> topicIds,
+        SubscribedTopicDescriber subscribedTopicDescriber
+    ) {
+        return topicIds.stream()
+            .flatMap(topic ->
+                IntStream.range(0, subscribedTopicDescriber.numPartitions(topic))
+                .mapToObj(i -> new TopicIdPartition(topic, i))
```

Comment:
nit: The `mapToObj` seems misplaced. How about?
```
IntStream
     .range(0, subscribedTopicDescriber.numPartitions(topic))
     .mapToObj(i -> new TopicIdPartition(topic, i))
```

Reply:
oh yep changed it, thanks!